import "Ws.proto";
package Magma;

option java_package = "org.obiba.opal.web.model";

message DatasourceDto {
  required string name = 1;
  optional string link = 2;
  repeated string table = 3;
}

message TableDto {
  required string name = 1;
  required string entityType = 2;
  optional string link = 3;
  optional int32 variableCount = 4;
  optional int32 valueSetCount = 5;
  optional string datasourceName = 6;
}

message VariableDto {
  required string name = 1;
  required string entityType = 2;
  required string valueType = 3;
  optional string mimeType = 4;
  required bool isRepeatable = 5;
  optional string occurrenceGroup = 6;  
  optional string link = 7;
  repeated AttributeDto attributes = 8;
  repeated CategoryDto categories = 9;
  optional string unit = 10;
  optional LinkDto parentLink = 11;
}

message CategoryDto {
  required string name = 1;
  required bool isMissing = 2;
  repeated AttributeDto attributes = 3;
}

message AttributeDto {
  required string name = 1;
  required string value = 2;
  optional string locale = 3;
}

message FrequencyDto {
  required string name = 1;
  required int32 value = 2;
}

message LinkDto {
  required string rel = 1;
  required string link = 2;
}

message ValueDto {
  required string valueType = 1;
  required bool isSequence = 2;
  optional string value = 3;
}

message DescriptiveStatsDto {
  optional double min = 1;
  optional double max = 2;
  optional double mean = 3;
  optional double geometricMean = 4;
  optional int64 n = 5;
  optional double sum = 6;
  optional double sumsq = 7;
  optional double stdDev = 8;
  optional double variance = 9;
  repeated double percentiles = 10;
  optional double skewness = 11;
  optional double kurtosis = 12;
  optional double median = 13;
  repeated double values = 14;
  repeated double distributionPercentiles = 15;
}
